---
title: "PGC Internship 2024: R Team - 3rd Progress Report"
author: "[Mentor] Mark Joseph Garrovillas, RCh, Leda Celeste Samin, Tyrone Mariano"
date: last-modified
institution: "Core Facility for Bioinformatics"
format: 
  revealjs:
    date-format: long
    slide-number: c/t
    width: 1920
    height: 1080
    theme: custom.scss
    navigation-mode: vertical
    controls-layout: bottom-right
    controls-tutorial: true
    preview-links: auto
    slide-level: 3
    multiplex: true
    chalkboard: 
      boardmarker-width: 5
    search:
    transition: slide
    background-transition: fade
    transition-speed: slow
title-slide-attributes:
      data-background: progress-rep03_files/img/title-slide.png
      data-background-size: cover
      data-background-color: "#000"
      transition: slide
      background-transition: fade
      transition-speed: slow
---

# PROJECT RECAP {data-background="progress-rep03_files/img/topic-slide.png" transition="fade-in slide-out" transition-speed="slow"}

Develop and implement the bioinformatics functions, tools, widgets, or pipelines

1\. Deploy the RStudio, Shiny, and OpenCPU servers using Docker

\- Easy development and access: OpenCPU, Shiny, & R Studio can be accessed by the user in their browser using Docker

\- Application deployment: From git, an isolated standalone application can be started via Docker

2\. Expose as a stateless microservice using RESTful API and RPC the major steps in the scRNAseq pipeline and their visualizations

3\. Develop an interactive dashboard/reactive framework in Shiny for the scRNA pipeline


# ACCOMPLISHMENTS {data-background="progress-rep03_files/img/topic-slide.png" transition="fade-in slide-out" transition-speed="slow"}


## Using the Seurat scRNAseq preprocessing pipeline and visualize the data in lower dimensions (t-SNE) {data-background="progress-rep03_files/img/subtopic-slide-2-1.png" transition="fade-in slide-out" transition-speed="slow"}

```{r setup, echo=F}
library(Seurat)
library(tidyverse)
library(shiny)
library(DT)
library(crosstalk)
library(plotly)

pbmc <- readRDS("C:/Users/lcbsa/OneDrive/Documents/PGC Training/pgc_internship/pgc-dim-red/presentations/progress-report_03/progress-rep03_files/data/pbmc_tutorial.rds")
pbmc <- RunTSNE(pbmc, dims = 1:10, dim.embed = 3)

# Extract tSNE information from Seurat Object
tsne_1 <- pbmc[["tsne"]]@cell.embeddings[,1]
tsne_2 <- pbmc[["tsne"]]@cell.embeddings[,2]
tsne_3 <- pbmc[["tsne"]]@cell.embeddings[,3]

# Prepare a dataframe for cell plotting
plot.data <- FetchData(object = pbmc, vars = c("tSNE_1", "tSNE_2", "tSNE_3", "seurat_clusters"))

# Make a column of row name identities (these will be your cell/barcode names)
plot.data$label <- paste(rownames(plot.data))

plot <- plot_ly(data = plot.data, x = ~tSNE_1, y = ~tSNE_2, z = ~tSNE_3, 
                color = ~seurat_clusters,
                type = "scatter3d",
                mode = "markers",
                marker = list(size = 5, width=2), # controls size of points
                text=~label, #This is that extra column we made earlier for which we will use 
                hoverinfo="text") #When you visualize your plotly object, hovering your mouse pointer over a point shows cell names

plot <- plot %>% layout(width = "100%", height = 500)

plot
```


## Creation of Docker images and containers {data-background="progress-rep03_files/img/subtopic-slide-2-1.png" transition="fade-in slide-out" transition-speed="slow"}
![](progress-rep03_files/img/recipe.png){.fragment fig-align="center" height="100%" width="100%"}

## Successful deployment of RStudio, Shiny, and OpenCPU servers using Docker {data-background="progress-rep03_files/img/subtopic-slide-2-1.png" transition="fade-in slide-out" transition-speed="slow"}

![](progress-rep03_files/img/runningimages.png){.fragment fig-align="center" height="100%" width="100%"}

## Access to servers via Docker {data-background="progress-rep03_files/img/subtopic-slide-2-1.png" transition="fade-in slide-out" transition-speed="slow"}

![Access to RStudio via web browser](progress-rep03_files/img/rstudioserver.png){.fragment fig-align="center" height="100%" width="100%"}

## Access to servers via Docker {data-background="progress-rep03_files/img/subtopic-slide-2-1.png" transition="fade-in slide-out" transition-speed="slow"}

![Access to OpenCPU via web browser](progress-rep03_files/img/opencpuweb.png){.fragment fig-align="center" height="100%" width="100%"}

## RPC major steps in scRNAseq pipeline via OpenCPU using Docker: Running the pipeline {data-background="progress-rep03_files/img/subtopic-slide-2-1.png" transition="fade-in slide-out" transition-speed="slow"}

![](progress-rep03_files/img/overviewpipeline.png){.fragment fig-align="center" height="100%" width="100%"}

## Developed packages which were uploaded in github {data-background="progress-rep03_files/img/subtopic-slide-2-1.png" transition="fade-in slide-out" transition-speed="slow"}

![](progress-rep03_files/img/packages.png){.fragment fig-align="center" height="100%" width="100%"}

## Creating the Seurat Object {data-background="progress-rep03_files/img/subtopic-slide-2-1.png" transition="fade-in slide-out" transition-speed="slow"}

![](progress-rep03_files/img/createseuratobject.png){.fragment fig-align="center" height="100%" width="100%"}

## Output: Seurat object in .rds format {data-background="progress-rep03_files/img/subtopic-slide-2-1.png" transition="fade-in slide-out" transition-speed="slow"}

![](progress-rep03_files/img/create_output.png){.fragment fig-align="center" height="100%" width="100%"}

## STEP 1: Quality Control {data-background="progress-rep03_files/img/subtopic-slide-2-1.png" transition="fade-in slide-out" transition-speed="slow"}

![](progress-rep03_files/img/qualitycontrol.png){.fragment fig-align="center" height="100%" width="100%"}

## Output: Seurat object after QC in .rds format {data-background="progress-rep03_files/img/subtopic-slide-2-1.png" transition="fade-in slide-out" transition-speed="slow"}

![](progress-rep03_files/img/qc_output.png){.fragment fig-align="center" height="100%" width="100%"}

## STEP 1: Visualize the Metrices {data-background="progress-rep03_files/img/subtopic-slide-2-1.png" transition="fade-in slide-out" transition-speed="slow"}

![](progress-rep03_files/img/metrices.png){.fragment fig-align="center" height="100%" width="100%"}

## Output: 2D interactive violin plots (html and png) {data-background="progress-rep03_files/img/subtopic-slide-2-1.png" transition="fade-in slide-out" transition-speed="slow"}

![](progress-rep03_files/img/metrices_output.png){.fragment fig-align="center" height="100%" width="100%"}

## STEP 2: Normalization {data-background="progress-rep03_files/img/subtopic-slide-2-1.png" transition="fade-in slide-out" transition-speed="slow"}

![](progress-rep03_files/img/normalization.png){.fragment fig-align="center" height="100%" width="100%"}

## Output: Normalized Seurat Object {data-background="progress-rep03_files/img/subtopic-slide-2-1.png" transition="fade-in slide-out" transition-speed="slow"}

![](progress-rep03_files/img/norm_output.png){.fragment fig-align="center" height="100%" width="100%"}


## STEP 3: Identification of Variable Features {data-background="progress-rep03_files/img/subtopic-slide-2-1.png" transition="fade-in slide-out" transition-speed="slow"}

![](progress-rep03_files/img/variable.png){.fragment fig-align="center" height="100%" width="100%"}


## Ouput: Identification of Variable Features {data-background="progress-rep03_files/img/subtopic-slide-2-1.png" transition="fade-in slide-out" transition-speed="slow"}

![](progress-rep03_files/img/features.png){.fragment fig-align="center" height="100%" width="100%"}


## Flowchart on the use of the packages for the scRNAseq pipeline {data-background="progress-rep03_files/img/subtopic-slide-2-1.png" transition="fade-in slide-out" transition-speed="slow"}
![](progress-rep03_files/img/flow.png){.fragment fig-align="center" height="100%" width="100%"}



## STEP 5: PCA {data-background="progress-rep03_files/img/subtopic-slide-2-1.png" transition="fade-in slide-out" transition-speed="slow"}

![](progress-rep03_files/img/pca.png){.fragment fig-align="center" height="100%" width="100%"}


## Output: PCA Plot {data-background="progress-rep03_files/img/subtopic-slide-2-1.png" transition="fade-in slide-out" transition-speed="slow"}

![](progress-rep03_files/img/pca_output.png){.fragment fig-align="center" height="100%" width="100%"}


## STEP 7: Non linear dimensionality reduction via t-SNE {data-background="progress-rep03_files/img/subtopic-slide-2-1.png" transition="fade-in slide-out" transition-speed="slow"}

![](progress-rep03_files/img/tsne.png){.fragment fig-align="center" height="100%" width="100%"}


## Output: t-SNE plot {data-background="progress-rep03_files/img/subtopic-slide-2-1.png" transition="fade-in slide-out" transition-speed="slow"}

![](progress-rep03_files/img/tsne_output.png){.fragment fig-align="center" height="100%" width="100%"}

# CHALLENGES {data-background="progress-rep03_files/img/topic-slide.png" transition="fade-in slide-out" transition-speed="slow"}

-   Challenges in writing and troubleshooting scripts
-   Familiarization with Docker via "doing" which included a lot of mistakes


# FOR NEXT WEEK {data-background="progress-rep03_files/img/topic-slide.png" transition="fade-in slide-out" transition-speed="slow"}

-   Develop more descriptive package and function names
-   Improve the script functionality
-   Create API documentation
-   Deploy a working pipeline to shiny servers via Docker


## Burndown Chart {data-background="progress-rep03_files/img/subtopic-slide-2-1.png" transition="fade-in slide-out" transition-speed="slow"}
![](progress-rep03_files/img/burndownchart.png){.fragment fig-align="center" height="100%" width="100%"}


## Burndown Chart {data-background="progress-rep03_files/img/subtopic-slide-2-1.png" transition="fade-in slide-out" transition-speed="slow"}
![](progress-rep03_files/img/burndowngraph.png){.fragment fig-align="center" height="100%" width="100%"}



## REFERENCES {data-background="progress-rep03_files/img/subtopic-slide-2-1.png" transition="fade-in slide-out" transition-speed="slow"}

-   Journal Articles:

    -   <https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9148156/>

-   Web Articles:

    -   <https://pmirla.github.io/archive/opencpu/index.html>
    -   \<https://www.bioconductor.org/packages/release/workflows/vignettes/sequencing/inst/doc/sequencing.htm

# Thank You! {data-background="progress-rep03_files/img/subtopic-slide-2-1.png" transition="fade-in slide-out" transition-speed="slow"}
